packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1228 @@
+  /// the [Router] uses this delegate to parse the route information and produce a
+  /// configuration. The configuration will be used by [routerDelegate] and
+  /// eventually rebuilds the [Router] widget.
+  ///
+  /// Since this delegate is the primary consumer of the [routeInformationProvider],
+  /// it must not be null if [routeInformationProvider] is not null.
+  final RouteInformationParser<T> routeInformationParser;
+
+  /// The router delegate for the router.
+  ///
+  /// This delegate consumes the configuration from [routeInformationParser] and
+  /// builds a navigating widget for the [Router].
+  ///
+  /// It is also the primary respondent for the [backButtonDispatcher]. The
+  /// [Router] relies on the [RouterDelegate.popRoute] to handles the back
+  /// button intends.
+  ///
+  /// If the [RouterDelegate.currentConfiguration] returns a non-null object,
+  /// this [Router] will opt for URL updates.
+  final RouterDelegate<T> routerDelegate;
+
+  /// The back button dispatcher for the router.
+  ///
+  /// The two common alternatives are the [RootBackButtonDispatcher] for root
+  /// router, or the [ChildBackButtonDispatcher] for other routers.
+  final BackButtonDispatcher backButtonDispatcher;
+
+  /// Retrieves the immediate [Router] ancestor from the given context.
+  ///
+  /// Use this method when you need to access the delegates in the [Router].
+  /// For example, you need to access the [backButtonDispatcher] of the parent
+  /// router to create a [ChildBackButtonDispatcher] for a nested router.
+  /// Another use case may be updating the value in [routeInformationProvider]
+  /// to navigate to a new route.
+  static Router<dynamic> of(BuildContext context) {
+    final _RouterScope scope = context.dependOnInheritedWidgetOfExactType<_RouterScope>();
+    assert(scope != null);
+    return scope.routerState.widget;
+  }
+
+  /// Forces the [Router] to to run the [callback] and reports the route
+  /// information back to the engine.
+  ///
+  /// Normally, calls to this method are not necessary because the [Router] will
+  /// reports route information back to the engine automatically if it changes.
+  /// Directly run the content of the [callback] should be sufficient. Only use
+  /// this method when you want to force the [Router] to reports the route
+  /// information regardless after running the [callback].
+  ///
+  /// See also:
+  ///
+  ///  * [Router]: see the "URL updates for web applications" section for more
+  ///    information about route information reporting.
+  static void navigate(BuildContext context, VoidCallback callback) {
+    final _RouterScope scope = context.getElementForInheritedWidgetOfExactType<_RouterScope>().widget as _RouterScope;
+    scope.routerState._setStateWithExplicitReportStatus(
+      _RouteInformationReportingStatus.must,
+      callback
+    );
+  }
+
+  /// Forces the [Router] to to run the [callback] without reporting the route
```

Comment:
When would I want to do this as an app developer?

Reply:
And what happens if I do `navigate` and `neglect` in the same frame?

Reply:
That means two changes happened: one that requires navigation, and one that doesn't. I think navigation should happen in this case.

This would be a good addition to the docs.